include/stdexec/execution.hpp

```diff
@@ -1377,29 +1377,15 @@ namespace stdexec {
         using __result_t = __await_result_t<_Awaitable, __promise_t<_Receiver>>;
         std::exception_ptr __eptr;
         try {
-          // This is a bit mind bending control-flow wise.
-          // We are first evaluating the co_await expression.
-          // Then the result of that is passed into a lambda
-          // that curries a reference to the result into another
-          // lambda which is then returned to 'co_yield'.
-          // The 'co_yield' expression then invokes this lambda
-          // after the coroutine is suspended so that it is safe
-          // for the receiver to destroy the coroutine.
-          auto __fun = [&](auto&&... __as) noexcept {
-            return [&]() noexcept -> void {
-              set_value((_Receiver&&) __rcvr, (std::add_rvalue_reference_t<__result_t>) __as...);
-            };
-          };
           if constexpr (std::is_void_v<__result_t>)
-            co_yield (co_await (_Awaitable&&) __await, __fun());
+            co_await (co_await (_Awaitable&&) __await, __co_call(set_value, (_Receiver&&) __rcvr));
           else
-            co_yield __fun(co_await (_Awaitable&&) __await);
+            co_await __co_call(set_value, (_Receiver&&) __rcvr, co_await (_Awaitable&&) __await);
+
         } catch (...) {
           __eptr = std::current_exception();
         }
-        co_yield [&]() noexcept -> void {
-          set_error((_Receiver&&) __rcvr, (std::exception_ptr&&) __eptr);
-        };
+        co_await __co_call(set_error, (_Receiver&&) __rcvr, (std::exception_ptr&&) __eptr);
```

Comment:
I wonder if this call (and the whole try/catch + `__eptr` variable could be conditionally avoided if the `co_await` expression was noexcept.
It could avoid having to add `exception_ptr` as an error completion signature in that case.

Reply:
How can I portably detect that?

Reply:
Good question. Ideally there would be a `__get_awaiter()` helper that called `operator co_await()` for you and that you could detect the noexceptness of that expression and then just check the await_ready/await_resume/await_suspend methods for noexceptness.

Wrap it all up into a `is_nothrow_awaitable_v<A, P>` query.

Technically, it's not implementable portably in library at the moment, though, as library can't exactly mimic the overload resolution of `operator co_await()`. But you could get pretty close and then just handwave away something that requires compilers to provide the necessary intrinsics to make stdlib implementations do the right thing.

Reply:
Filed at #795